docs/security/index.md

```diff
@@ -0,0 +1,453 @@
+encode labels as [`xsd:base64Binary`](XsdBase64) typed literals e.g.
+
+```ttl
+@prefix xsd:     <http://www.w3.org/2001/XMLSchema#> .
+@prefix foaf:    <http://xmlns.com/foaf/0.1/> .
+@prefix authz:   <http://telicent.io/security#> .
+@prefix data:    <http://data.gov.uk/testdata#> .
+
+[ authz:pattern 'data:Fred foaf:age 34' ; authz:label "Z2Rwcg=="^^xsd:base64Binary ] .
+```
+
+This allows us to encode arbitrary label byte sequences in a string based syntax using the standard RDF datatype
+mechanics.
+
+Again for backwards compatibility existing RDF-ABAC labels graphs, where labels are given as plain string literals, are
+still honoured, so the following existing valid labels graph would continue to be considered equivalent:
+
+```ttl
+@prefix xsd:     <http://www.w3.org/2001/XMLSchema#> .
+@prefix foaf:    <http://xmlns.com/foaf/0.1/> .
+@prefix authz:   <http://telicent.io/security#> .
+@prefix data:    <http://data.gov.uk/testdata#> .
+
+[ authz:pattern 'data:Fred foaf:age 34' ; authz:label "gdpr" ] .
+```
+
+### Entitlements
+
+Our existing security model closely couples RDF-ABAC labels with the User Attributes returned by the [Telicent
+Access][TcAccess] service where user attributes are returned in a specific JSON object schema.  Applications are also
+strongly coupled to Access in that they need to be directly aware of configuring themselves to talk to it in order to
+retrieve user attributes to provide label enforcement decisions.
+
+Therefore as part of this new API we adopt a similar approach to [Labels](#label-format--syntax) in that we move to
+treat user entitlements as opaque byte sequences with responsibility for retrieving and interpreting user entitlements
+left to plugins via their `EntitlementsProvider` and `EntitlementsParser` implementations.  This means that the burden
+of configuring any necessary supporting services, e.g. connectivity to Access, becomes an internal implementation detail
+of a plugin, rather than an application concern. Entitlements use the same optional schema prefix as part of the byte
+sequences to allow plugins to distinguish between different entitlement schemas they may support, and this allows for
+future plugins to potentially combine multiple entitlement providers.
+
+As with labels the `Entitlements<T>` interface treats entitlements in abstract terms from an application perspective
+while allowing a plugin implementation to decode the entitlements into whatever data structure(s) are appropriate. A
+plugins implementation of the `EntitlementsParser` interface can supply whatever decoding logic it needs for its
+implementation.
+
+### Data Access Enforcement
+
+Data Access Enforcement is done by combining the users retrieved [Entitlements](#entitlements) with the
+[Labels](#label-format--syntax) present on the data.  This is done by first preparing an `Authorizer` via the
+`prepareAuthorizer()` method passing in the users entitlements.  `Authorizer` instances are intended to be scoped to the
```

Comment:
Could this be expanded to include some flexible context as well as the user's entitlement?

For example ODY have suggested policies/conditions based on location and device authentication. These attributes will likely need to come from the connection / request somehow.

Reply:
The thinking, which probably wasn't clear enough here, is that the entitlements information that gets retrieved would include that kind of detail.

Of course it may be that the interfaces as defined in this 1st draft aren't sufficient for that, e.g. perhaps `prepareAuthorizer()` needs to take an additional parameter which provides access to the connection/request details somehow